registry/remote/auth/scope.go

```diff
@@ -93,7 +119,62 @@ func AppendScopes(ctx context.Context, scopes ...string) context.Context {
 // GetScopes returns the scopes in the context.
 func GetScopes(ctx context.Context) []string {
 	if scopes, ok := ctx.Value(scopesContextKey{}).([]string); ok {
-		return append([]string(nil), scopes...)
+		return slices.Clone(scopes)
+	}
+	return nil
+}
+
+// scopesPerHostContextKey is the context key for per-host scopes.
+type scopesPerHostContextKey struct{}
+
+// WithScopesPerHost returns a context with per-host scopes added.
+// Scopes are de-duplicated.
+// Scopes are used as hints for the auth client to fetch bearer tokens with
+// larger scopes.
+//
+// For example, uploading blob to the repository "hello-world" does HEAD request
+// first then POST and PUT. The HEAD request will return a challenge for scope
+// `repository:hello-world:pull`, and the auth client will fetch a token for
+// that challenge. Later, the POST request will return a challenge for scope
+// `repository:hello-world:push`, and the auth client will fetch a token for
+// that challenge again. By invoking WithScopesPerHost with the scope
+// `repository:hello-world:pull,push`, the auth client with cache is hinted to
+// fetch a token via a single token fetch request for all the HEAD, POST, PUT
+// requests.
+//
+// Passing an empty list of scopes will virtually remove the scope hints in the
+// context for the given host.
+//
+// Reference: https://docs.docker.com/registry/spec/auth/scope/
+func WithScopesPerHost(ctx context.Context, host string, scopes ...string) context.Context {
+	var scopesByHost map[string][]string
+	if old, ok := ctx.Value(scopesPerHostContextKey{}).(map[string][]string); ok {
+		scopesByHost = make(map[string][]string, len(old))
+		maps.Copy(scopesByHost, old)
+	} else {
+		scopesByHost = make(map[string][]string, 1)
+	}
+
+	scopesByHost[host] = CleanScopes(scopes)
```

Comment:
Is there going to be a data race condition, if more than one threads simultaneously reach L153 and L158?

Reply:
Good catch! There is a Time-of-Check to Time-of-Use race condition... I'm still finding a way to fix it.

Reply:
Updated the code and this is not valid anymore.